src/frequenz/sdk/timeseries/_formula_engine/_formula_formatter.py

```diff
@@ -0,0 +1,272 @@
+                case Multiplier():
+                    self._format_binary(Operator.MULTIPLICATION)
+                case Divider():
+                    self._format_binary(Operator.DIVISION)
+                case Averager():
+                    value = (
+                        # pylint: disable=protected-access
+                        f"avg({', '.join(self._format([f]) for f in step.fetchers)})"
+                    )
+                    self._stack.append(StackItem(value, OperatorPrecedence.PRIMARY, 1))
+                case Clipper():
+                    the_value = self._stack.pop()
+                    min_value = step.min_value if step.min_value is not None else "-inf"
+                    max_value = step.max_value if step.max_value is not None else "inf"
+                    value = f"clip({min_value}, {the_value.value}, {max_value})"
+                    self._stack.append(StackItem(value, OperatorPrecedence.PRIMARY, 1))
+                case Maximizer():
+                    left, right = self._pop_two_from_stack()
+                    value = f"max({left.value}, {right.value})"
+                    self._stack.append(StackItem(value, OperatorPrecedence.PRIMARY, 1))
+                case Minimizer():
+                    left, right = self._pop_two_from_stack()
+                    value = f"min({left.value}, {right.value})"
+                    self._stack.append(StackItem(value, OperatorPrecedence.PRIMARY, 1))
+                case MetricFetcher():
+                    metric_fetcher = step
+                    value = metric_fetcher._name  # pylint: disable=protected-access
+                    if engine_reference := getattr(
+                        metric_fetcher.stream, "_engine_reference", None
+                    ):
+                        value = f"[{value}]({str(engine_reference)})"
+                    self._stack.append(StackItem(value, OperatorPrecedence.PRIMARY, 1))
+                case OpenParen():
+                    pass  # We gently ignore this one.
+
+        assert (
+            len(self._stack) == 1
+        ), f"The formula {postfix_expr} is not valid. Evaluation stack left-over: {self._stack}"
+        return self._stack[0].value
+
+    def _format_binary(self, operator: Operator) -> None:
+        """Format a binary operation.
+
+        Pops the arguments of the binary expression from the stack
+        and pushes the string representation of the binary operation to the stack.
+
+        Args:
+            operator: The operator of the binary operation.
+        """
+        left, right = self._pop_two_from_stack()
+        self._stack.append(StackItem.create_binary(left, operator, right))
+
+    def _pop_two_from_stack(self) -> tuple[StackItem, StackItem]:
+        """Pop two items from the stack.
+
+        Returns:
+            The two items popped from the stack.
+        """
+        right = self._stack.pop()
+        left = self._stack.pop()
+        return left, right
```

Comment:
```suggestion
        return left, right


def format_formula(postfix_expr: list[FormulaStep]) -> str:
        """Return a representation of the formula as a string in infix notation.

        Args:
            postfix_expr: The steps of the formula in postfix notation order.

        Returns:
            str: The formula in infix notation.
        """
        formatter = FormulaFormatter()
        return formatter._format(postfix_expr)
```

Not using a global since it was 2 against 1 :P